File: examples/csharp/Fulfillment/src/FulfillmentLogic.cs
using Angzarr;
using Examples;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Serilog;

namespace Angzarr.Examples.Fulfillment;

public class FulfillmentLogic : IFulfillmentLogic
{
    private readonly Serilog.ILogger _logger;

    public FulfillmentLogic(Serilog.ILogger logger)
    {
        _logger = logger.ForContext<FulfillmentLogic>();
    }

    public FulfillmentState RebuildState(EventBook? eventBook)
    {
        var state = FulfillmentState.Empty;

        if (eventBook == null || eventBook.Pages.Count == 0)
            return state;

        foreach (var page in eventBook.Pages)
        {
            if (page.Event == null) continue;
            state = ApplyEvent(state, page.Event);
        }

        return state;
    }

    private FulfillmentState ApplyEvent(FulfillmentState state, Any eventAny)
    {
        var typeUrl = eventAny.TypeUrl;

        if (typeUrl.EndsWith("ShipmentCreated"))
        {
            var evt = eventAny.Unpack<ShipmentCreated>();
            return state with { OrderId = evt.OrderId, Status = "pending" };
        }

        if (typeUrl.EndsWith("ItemsPicked"))
        {
            return state with { Status = "picking" };
        }

        if (typeUrl.EndsWith("ItemsPacked"))
        {
            return state with { Status = "packing" };
        }

        if (typeUrl.EndsWith("Shipped"))
        {
            var evt = eventAny.Unpack<Shipped>();
            return state with { Status = "shipped", TrackingNumber = evt.TrackingNumber };
        }

        if (typeUrl.EndsWith("Delivered"))
        {
            return state with { Status = "delivered" };
        }

        return state;
    }

    public EventBook HandleCreateShipment(FulfillmentState state, string orderId)
    {
        if (state.Exists)
            throw CommandValidationException.FailedPrecondition("Shipment already exists");

        if (string.IsNullOrWhiteSpace(orderId))
            throw CommandValidationException.InvalidArgument("Order ID is required");

        _logger.Information("creating_shipment {@Data}", new { order_id = orderId });

        var evt = new ShipmentCreated
        {
            OrderId = orderId,
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleMarkPicked(FulfillmentState state)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Shipment does not exist");

        if (!state.IsPending)
            throw CommandValidationException.FailedPrecondition("Shipment not in pending state");

        _logger.Information("marking_picked {@Data}", new { order_id = state.OrderId });

        var evt = new ItemsPicked
        {
            PickedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleMarkPacked(FulfillmentState state)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Shipment does not exist");

        if (!state.IsPicking)
            throw CommandValidationException.FailedPrecondition("Items not picked yet");

        _logger.Information("marking_packed {@Data}", new { order_id = state.OrderId });

        var evt = new ItemsPacked
        {
            PackedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleShip(FulfillmentState state, string trackingNumber, string carrier)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Shipment does not exist");

        if (!state.IsPacking)
            throw CommandValidationException.FailedPrecondition("Items not packed yet");

        if (string.IsNullOrWhiteSpace(trackingNumber))
            throw CommandValidationException.InvalidArgument("Tracking number is required");

        _logger.Information("shipping {@Data}", new { order_id = state.OrderId, tracking = trackingNumber, carrier });

        var evt = new Shipped
        {
            TrackingNumber = trackingNumber,
            Carrier = carrier ?? "",
            ShippedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleRecordDelivery(FulfillmentState state)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Shipment does not exist");

        if (!state.IsShipped)
            throw CommandValidationException.FailedPrecondition("Order not shipped yet");

        _logger.Information("recording_delivery {@Data}", new { order_id = state.OrderId });

        var evt = new Delivered
        {
            DeliveredAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    private static EventBook CreateEventBook(IMessage evt)
    {
        var page = new EventPage
        {
            Num = 0,
            Event = Any.Pack(evt),
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        var book = new EventBook();
        book.Pages.Add(page);
        return book;
    }
}

File: examples/python/fulfillment/fulfillment_logic.py
"""Fulfillment command handlers."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains

from state import FulfillmentState


class CommandRejectedError(Exception):
    """Command was rejected due to business rule violation."""


def handle_create_shipment(command_book, command_any, state: FulfillmentState, seq: int, log) -> angzarr.EventBook:
    if state.exists():
        raise CommandRejectedError("Shipment already exists")

    cmd = domains.CreateShipment()
    command_any.Unpack(cmd)

    if not cmd.order_id:
        raise CommandRejectedError("Order ID is required")

    log.info("creating_shipment", order_id=cmd.order_id)

    event = domains.ShipmentCreated(
        order_id=cmd.order_id,
        status="pending",
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_mark_picked(command_book, command_any, state: FulfillmentState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Shipment does not exist")
    if not state.is_pending():
        raise CommandRejectedError(f"Cannot pick items in {state.status} state")

    cmd = domains.MarkPicked()
    command_any.Unpack(cmd)

    if not cmd.picker_id:
        raise CommandRejectedError("Picker ID is required")

    log.info("marking_picked", picker_id=cmd.picker_id)

    event = domains.ItemsPicked(
        picker_id=cmd.picker_id,
        picked_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_mark_packed(command_book, command_any, state: FulfillmentState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Shipment does not exist")
    if not state.is_picking():
        raise CommandRejectedError(f"Cannot pack items in {state.status} state")

    cmd = domains.MarkPacked()
    command_any.Unpack(cmd)

    if not cmd.packer_id:
        raise CommandRejectedError("Packer ID is required")

    log.info("marking_packed", packer_id=cmd.packer_id)

    event = domains.ItemsPacked(
        packer_id=cmd.packer_id,
        packed_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_ship(command_book, command_any, state: FulfillmentState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Shipment does not exist")
    if not state.is_packing():
        raise CommandRejectedError(f"Cannot ship in {state.status} state")

    cmd = domains.Ship()
    command_any.Unpack(cmd)

    if not cmd.carrier:
        raise CommandRejectedError("Carrier is required")
    if not cmd.tracking_number:
        raise CommandRejectedError("Tracking number is required")

    log.info("shipping", carrier=cmd.carrier, tracking_number=cmd.tracking_number)

    event = domains.Shipped(
        carrier=cmd.carrier,
        tracking_number=cmd.tracking_number,
        shipped_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )


def handle_record_delivery(command_book, command_any, state: FulfillmentState, seq: int, log) -> angzarr.EventBook:
    if not state.exists():
        raise CommandRejectedError("Shipment does not exist")
    if not state.is_shipped():
        raise CommandRejectedError(f"Cannot record delivery in {state.status} state")

    cmd = domains.RecordDelivery()
    command_any.Unpack(cmd)

    log.info("recording_delivery", signature=cmd.signature)

    event = domains.Delivered(
        signature=cmd.signature,
        delivered_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )
